clustermesh/clustermesh.go

```diff
@@ -494,7 +503,9 @@ func NewK8sClusterMesh(client k8sClusterMeshImplementation, p Parameters) *K8sCl
 }
 
 func (k *K8sClusterMesh) Log(format string, a ...interface{}) {
-	fmt.Fprintf(k.params.Writer, format+"\n", a...)
+	if k.params.Output == OutputSummary {
```

Comment:
This seems like the wrong place to put this conditional.

IMHO it would make more sense for the status command to compute the objects, then either output json or summary as configured.

Reply:
I agree in principle. But the status is actually printed bit by bit at the moment, using this `Log` method.
In order to support JSON output, I need to:
- either make this method silent
- or build up all status strings as an array and print it at the end, which is a much larger refactoring of the summary logic

Reply:
I've fixed this up as follows:

```diff
diff --git a/clustermesh/clustermesh.go b/clustermesh/clustermesh.go
index 2095c90565e9..9f30afd39501 100644
--- a/clustermesh/clustermesh.go
+++ b/clustermesh/clustermesh.go
@@ -500,9 +500,7 @@ func NewK8sClusterMesh(client k8sClusterMeshImplementation, p Parameters) *K8sCl
 }
 
 func (k *K8sClusterMesh) Log(format string, a ...interface{}) {
-       if k.params.Output == status.OutputSummary {
-               fmt.Fprintf(k.params.Writer, format+"\n", a...)
-       }
+       fmt.Fprintf(k.params.Writer, format+"\n", a...)
 }
 
 func (k *K8sClusterMesh) GetClusterConfig(ctx context.Context) error {
diff --git a/internal/cli/cmd/clustermesh.go b/internal/cli/cmd/clustermesh.go
index dd394835a78a..9ebbdb6d4fb1 100644
--- a/internal/cli/cmd/clustermesh.go
+++ b/internal/cli/cmd/clustermesh.go
@@ -153,6 +153,12 @@ func newCmdClusterMeshStatus() *cobra.Command {
                RunE: func(cmd *cobra.Command, args []string) error {
                        params.Namespace = namespace
 
+                       if params.Output == status.OutputJSON {
+                               // Write status log messages to stderr to make sure they don't
+                               // clutter JSON output.
+                               params.Writer = os.Stderr
+                       }
+
                        cm := clustermesh.NewK8sClusterMesh(k8sClient, params)
                        if _, err := cm.Status(context.Background()); err != nil {
                                fatalf("Unable to determine status:  %s", err)
```